crates/algokit_transact/src/transactions/application_call.rs

```diff
@@ -0,0 +1,331 @@
+    /// This cannot be changed after creation.
+    #[serde(rename = "apep")]
+    #[serde(skip_serializing_if = "is_zero_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub extra_program_pages: Option<u64>,
+
+    /// Transaction specific arguments available in the application's
+    /// approval program and clear state program.
+    #[serde(rename = "apaa")]
+    #[serde_as(as = "Option<Vec<Bytes>>")]
+    #[serde(skip_serializing_if = "is_empty_vec_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub args: Option<Vec<Vec<u8>>>,
+
+    /// List of accounts in addition to the sender that may be accessed
+    /// from the application's approval program and clear state program.
+    #[serde(rename = "apat")]
+    #[serde(skip_serializing_if = "is_empty_vec_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub account_references: Option<Vec<Address>>,
+
+    /// List of applications in addition to the application ID that may be called
+    /// from the application's approval program and clear state program.
+    #[serde(rename = "apfa")]
+    #[serde(skip_serializing_if = "is_empty_vec_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub app_references: Option<Vec<u64>>,
+
+    /// Lists the assets whose parameters may be accessed by this application's
+    /// approval program and clear state program.
+    ///
+    /// The access is read-only.
+    #[serde(rename = "apas")]
+    #[serde(skip_serializing_if = "is_empty_vec_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub asset_references: Option<Vec<u64>>,
+
+    /// The boxes that should be made available for the runtime of the program.
+    #[serde(rename = "apbx")]
+    #[serde(skip_serializing_if = "is_empty_vec_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub box_references: Option<Vec<BoxReference>>,
+}
+
+fn is_default_on_complete(on_complete: &OnApplicationComplete) -> bool {
+    matches!(on_complete, OnApplicationComplete::NoOp)
+}
+
+/// Custom serializer for application call transactions.
+///
+/// This serializer handles the special case of box references, where app IDs need to be
+/// transformed from actual application IDs to positional indices for wire format compatibility.
+pub fn application_call_serializer<S>(
+    fields: &ApplicationCallTransactionFields,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    // Transform box references if present
+    if let Some(ref box_references) = fields.box_references {
+        if !box_references.is_empty() {
+            let app_references = fields.app_references.as_deref().unwrap_or(&[]);
+
+            let box_references = box_references
+                .iter()
+                .map(|box_ref| {
+                    let app_id_index = if box_ref.app_id == 0 || box_ref.app_id == fields.app_id {
+                        // App ID 0 and the transaction's own app ID remain as 0 (current application)
```

Comment:
I think the comment can be a bit clearer
```suggestion
                        // Return 0 if the App ID is 0 or the App ID is the same with the current application ID
```
something along this line, what do you think?